pkg/controller/infra/infra_controller.go

```diff
@@ -525,101 +486,68 @@ func (c *InfraController) syncCluster(key string) error {
 
 	cluster, err := c.clustersLister.Clusters(ns).Get(name)
 	if errors.IsNotFound(err) {
-		cLog.Debugln("cluster not found")
-		c.expectations.DeleteExpectations(key)
+		cLog.Debugln("cluster deleted")
+		c.jobControl.ObserveOwnerDeletion(key)
 		return nil
 	}
 	if err != nil {
 		return err
 	}
 
-	if !c.expectations.SatisfiedExpectations(key) {
-		// expectations have not been met, come back later
-		cLog.Debugln("cluster has not satisfied expectations yet")
-		return nil
-	}
+	specChanged := cluster.Status.ProvisionedJobGeneration != cluster.Generation
+
+	jobFactory := c.getJobFactory(cluster, specChanged)
```

Comment:
putting the logic in the job factory creation to decide whether we should create a job or not doesn't seem very explicit. I would rather pass a function to the JobControl interface that would decide, given an owner object, whether a new job is needed or not.

Reply:
I am hesitant to supply a function when creating the `JobControl` that would be used to determine whether a new job is needed. I could easily see cases where the decision is based on more than just information in the owner. Would it be OK if I pass an `isJobNeeded` boolean to `ControlJobs`?

Reply:
sgtm